File: samples/system/typescript/04_streaming.ts
/**
 * 04 - Streaming Chat Completion
 *
 * Demonstrates streaming responses with failover between OpenAI and Google
 * Gemini providers. This sample shows:
 *
 *   - Using the OpenAI client with stream=true for incremental output
 *   - Iterating over response chunks as they arrive
 *   - Automatic fallback to the next provider when a stream errors mid-flight
 *   - Delivery mode filtering (only stream-capable models are selected)
 *
 * Streaming is a delivery mode orthogonal to capabilities. When stream=true
 * is set, the routing pipeline filters for models that support streaming
 * delivery before applying the selection strategy.
 *
 * Prerequisites:
 *   - Set OPENAI_API_KEY and GOOGLE_API_KEY environment variables
 */

import { ModelMesh, MeshConfig } from "modelmesh-lite";

async function main(): Promise<void> {
  // -----------------------------------------------------------------------
  // 1. Configure two streaming-capable providers
  // -----------------------------------------------------------------------
  const config: MeshConfig = {
    raw: {
      providers: {
        "openai.llm.v1": {
          enabled: true,
          api_key: "${secrets:OPENAI_API_KEY}",
        },
        "google.gemini.v1": {
          enabled: true,
          api_key: "${secrets:GOOGLE_API_KEY}",
        },
      },

      pools: {
        "text-generation": {
          strategy: "modelmesh.stick-until-failure.v1",
          provider_priority: ["openai.llm.v1", "google.gemini.v1"],

          deactivation: {
            retry_limit: 2,
            error_codes: [429, 500, 502, 503],
          },
          recovery: {
            cooldown: "30s",
          },
          retry: {
            max_attempts: 1,
            backoff: "fixed",
            initial_delay: "500ms",
            retryable_codes: [429, 500, 502, 503],
            scope: "any",
          },
        },
      },

      observability: {
        routing: {
          connector: "modelmesh.console.v1",
        },
      },
    },
  };

  // -----------------------------------------------------------------------
  // 2. Initialize
  // -----------------------------------------------------------------------
  const mesh = new ModelMesh();
  await mesh.initialize(config);
  console.log("ModelMesh initialized with OpenAI + Gemini (streaming mode).\n");

  const client = mesh.getClient();

  // -----------------------------------------------------------------------
  // 3. Streaming chat completion
  // -----------------------------------------------------------------------
  console.log("--- Streaming Response ---");
  console.log("Prompt: Write a short story about a robot learning to paint.\n");

  // When stream=true, the client returns an AsyncIterable of chunks.
  // Each chunk contains incremental content from the selected model.
  const stream = await client.chat.completions.create({
    model: "text-generation",
    messages: [
      { role: "system", content: "You are a creative storyteller." },
      { role: "user", content: "Write a short story about a robot learning to paint. Keep it under 200 words." },
    ],
    temperature: 0.8,
    max_tokens: 300,
    stream: true,
  });

  // The stream object is an AsyncIterable. If the underlying provider
  // errors mid-stream, ModelMesh transparently retries or rotates to
  // the fallback provider and continues streaming.
  let fullContent = "";
  let chunkCount = 0;

  for await (const chunk of stream as AsyncIterable<any>) {
    const delta = chunk.choices?.[0]?.delta?.content ?? "";
    if (delta) {
      process.stdout.write(delta);
      fullContent += delta;
      chunkCount++;
    }
  }

  console.log("\n\n--- Stream Complete ---");
  console.log(`Total chunks received: ${chunkCount}`);
  console.log(`Total characters: ${fullContent.length}`);

  // -----------------------------------------------------------------------
  // 4. Second streaming request (demonstrates continued routing)
  // -----------------------------------------------------------------------
  console.log("\n--- Second Streaming Request ---");
  console.log("Prompt: Continue the story.\n");

  const stream2 = await client.chat.completions.create({
    model: "text-generation",
    messages: [
      { role: "system", content: "You are a creative storyteller." },
      { role: "user", content: "Write a short story about a robot learning to paint." },
      { role: "assistant", content: fullContent },
      { role: "user", content: "Continue the story in 100 words." },
    ],
    temperature: 0.8,
    max_tokens: 200,
    stream: true,
  });

  for await (const chunk of stream2 as AsyncIterable<any>) {
    const delta = chunk.choices?.[0]?.delta?.content ?? "";
    if (delta) {
      process.stdout.write(delta);
    }
  }

  console.log("\n\n--- Done ---");

  // -----------------------------------------------------------------------
  // 5. Shut down
  // -----------------------------------------------------------------------
  await mesh.shutdown();
  console.log("ModelMesh shut down.");
}

main().catch(console.error);

File: samples/system/python/04_streaming.py
"""
04 - Streaming Chat Completion
==============================

Streaming chat completion with automatic provider failover.  The response
arrives incrementally as the model generates tokens.

This sample demonstrates:
  - Using stream=True with the OpenAI-compatible client
  - Iterating over streaming response chunks
  - Automatic fallback to the next provider on stream error

Uses mock providers so it runs without API keys.
"""

import asyncio
from modelmesh import ModelMesh, MeshConfig
from modelmesh.interfaces.provider import (
    ChatMessage, CompletionChoice, CompletionRequest, CompletionResponse,
    ErrorClassification, ModelInfo, ModelPricing, ProviderConnector,
    QuotaStatus, RateLimitStatus, TokenUsage,
)


class MockStreamProvider(ProviderConnector):
    """Mock provider that yields streaming chunks."""

    def __init__(self, name: str, tokens: list[str]):
        self._name = name
        self._tokens = tokens

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        return CompletionResponse(
            id=f"resp-{self._name}",
            model=self._name,
            choices=[CompletionChoice(
                index=0,
                message=ChatMessage(role="assistant", content="".join(self._tokens)),
                finish_reason="stop",
            )],
            usage=TokenUsage(prompt_tokens=10, completion_tokens=len(self._tokens), total_tokens=10 + len(self._tokens)),
        )

    async def stream(self, request: CompletionRequest):
        for i, token in enumerate(self._tokens):
            yield CompletionResponse(
                id=f"chunk-{self._name}-{i}",
                model=self._name,
                choices=[CompletionChoice(
                    index=0,
                    delta=ChatMessage(role="assistant", content=token),
                    finish_reason="stop" if i == len(self._tokens) - 1 else None,
                )],
                usage=TokenUsage(prompt_tokens=10, completion_tokens=i + 1, total_tokens=11 + i),
            )

    def get_capabilities(self):
        return ["generation.text-generation.chat-completion"]

    def supports(self, cap):
        return cap in self.get_capabilities()

    def list_models(self):
        return [ModelInfo(id=self._name, name=self._name)]

    def get_model_info(self, mid):
        return ModelInfo(id=mid, name=mid)

    def check_quota(self):
        return QuotaStatus()

    def get_rate_limits(self):
        return RateLimitStatus()

    def get_pricing(self, mid):
        return ModelPricing()

    def report_usage(self, mid, usage):
        pass

    def classify_error(self, error):
        return ErrorClassification(retryable=False)


async def main() -> None:
    poem_tokens = [
        "In ", "the ", "network's ", "web ", "we ", "trust,\n",
        "Packets ", "flow ", "like ", "morning ", "dust.\n",
        "Consensus ", "reached ", "by ", "many ", "peers,\n",
        "Distributed ", "hope ", "conquers ", "fears.",
    ]
    list_tokens = [
        "1. ", "High ", "availability ", "-- ", "no ", "single ", "point ", "of ", "failure.\n",
        "2. ", "Scalability ", "-- ", "handle ", "more ", "traffic.\n",
        "3. ", "Performance ", "-- ", "reduced ", "latency.",
    ]

    prov_a = MockStreamProvider("gpt-4o", poem_tokens)
    prov_b = MockStreamProvider("gemini-flash", list_tokens)

    config = MeshConfig(raw={
        "providers": {
            "openai": {"connector": "openai", "enabled": True, "instance": prov_a},
            "google": {"connector": "google", "enabled": True, "instance": prov_b},
        },
        "models": {
            "openai.gpt-4o": {
                "provider": "openai",
                "capabilities": ["generation.text-generation.chat-completion"],
            },
            "google.gemini-flash": {
                "provider": "google",
                "capabilities": ["generation.text-generation.chat-completion"],
            },
        },
        "pools": {
            "text-generation": {
                "capability": "generation.text-generation.chat-completion",
                "strategy": "stick-until-failure",
            },
        },
    })
    mesh = ModelMesh()
    mesh.initialize(config)
    client = mesh.get_client()

    print("=" * 60)
    print("Streaming chat completion demo")
    print("=" * 60)

    # Example 1: Basic streaming
    print("\n--- Example 1: Basic streaming ---\n")
    print("User: Write a short poem about distributed systems.\n")
    print("Assistant: ", end="")

    response_stream = client.chat.completions.create(
        model="text-generation",
        messages=[
            {"role": "system", "content": "You are a creative writer."},
            {"role": "user", "content": "Write a short poem about distributed systems."},
        ],
        stream=True,
    )

    full_text = ""
    chunk = None
    for chunk in response_stream:
        if chunk.choices and chunk.choices[0].delta and chunk.choices[0].delta.content:
            token = chunk.choices[0].delta.content
            print(token, end="", flush=True)
            full_text += token

    print("\n")
    if chunk:
        print(f"Model used: {chunk.model}")

    # Force rotation to second model for the next stream demo
    mesh.rotate("text-generation")

    # Example 2: Streaming with metadata inspection
    print("\n--- Example 2: Streaming with usage metadata ---\n")
    print("User: List three benefits of load balancing.\n")
    print("Assistant: ", end="")

    chunk_count = 0
    response_stream = client.chat.completions.create(
        model="text-generation",
        messages=[{"role": "user", "content": "List three benefits of load balancing."}],
        stream=True,
    )

    chunk = None
    for chunk in response_stream:
        chunk_count += 1
        if chunk.choices and chunk.choices[0].delta and chunk.choices[0].delta.content:
            print(chunk.choices[0].delta.content, end="", flush=True)

    print(f"\n\nReceived {chunk_count} chunks.")
    if chunk:
        print(f"Final chunk usage: {chunk.usage}")

    mesh.shutdown()
    print("\nModelMesh shut down.")


if __name__ == "__main__":
    asyncio.run(main())
